File: BOJ/1074.java
import java.io.*;
import java.util.*;

public class Main {
	
	static int N, R, C, ans = 0;
	
	public static void main(String[] args) throws Exception {
		// System.setIn(new FileInputStream("res/input.txt"));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		R = Integer.parseInt(st.nextToken());
		C = Integer.parseInt(st.nextToken());
		
		solution(0, 0, (1<<N));
		br.close();
	}

	private static void solution(int r, int c, int size) {
		if(r==R && c==C) {
			System.out.println(ans);
			return ;
		}
		
		if(r<=R && R<r+size && c<=C && C<c+size) { // 범위 안에 있는 경우 분할 정복을 통해 검색 
			int half = size/2;
			solution(r, c, half);
			solution(r, c+half, half);
			solution(r+half, c, half);
			solution(r+half, c+half, half);
		}
		else ans+=size*size;
	}
}

File: BOJ/1074.cpp
#include<iostream>
using namespace std;

int N, R, C, ans=0;

void solution(int r, int c, int size) {
    if(r==R && c==C){
        cout << ans;
        return;
    }

    if(r<=R && R<r+size && c<=C && C<c+size){
        int half = size/2;
        solution(r, c, half); // 왼쪽 위칸
        solution(r, c+half, half); // 오른쪽 위칸
        solution(r+half, c, half); // 왼쪽 아래칸
        solution(r+half, c+half, half); // 오른쪽 아래칸
    }
    else ans+=size*size;
}

int main() {
	ios_base::sync_with_stdio(0); cin.tie(0);
	freopen("input.txt", "r", stdin); // task /f /PID [pid]
	cin >> N >> R >> C; 
	solution(0, 0, 1<<N);
}